mono/metadata/reflection.c

```
@@ -3808,97 +3900,125 @@ compare_types_by_table_idx (MonoReflectionTypeBuilder **type1, MonoReflectionTyp
 		return 0;
 }
 
-static void
-params_add_cattrs (MonoDynamicImage *assembly, MonoArray *pinfo) {
+static gboolean
+params_add_cattrs (MonoDynamicImage *assembly, MonoArray *pinfo, MonoError *error) {
+	gboolean succeeded = TRUE;
 	int i;
 
+	mono_error_init (error);
 	if (!pinfo)
-		return;
-	for (i = 0; i < mono_array_length (pinfo); ++i) {
+		return TRUE;
+	for (i = 0; succeeded && i < mono_array_length (pinfo); ++i) {
 		MonoReflectionParamBuilder *pb;
 		pb = mono_array_get (pinfo, MonoReflectionParamBuilder *, i);
 		if (!pb)
 			continue;
-		mono_image_add_cattrs (assembly, pb->table_idx, MONO_CUSTOM_ATTR_PARAMDEF, pb->cattrs);
+		succeeded = mono_image_add_cattrs (assembly, pb->table_idx, MONO_CUSTOM_ATTR_PARAMDEF, pb->cattrs, error);
 	}
+
+	return succeeded;
+
 }
 
-static void
-type_add_cattrs (MonoDynamicImage *assembly, MonoReflectionTypeBuilder *tb) {
+static gboolean
+type_add_cattrs (MonoDynamicImage *assembly, MonoReflectionTypeBuilder *tb, MonoError *error) {
+	gboolean succeeded = TRUE;
 	int i;
+
+	mono_error_init (error);
 	
-	mono_image_add_cattrs (assembly, tb->table_idx, MONO_CUSTOM_ATTR_TYPEDEF, tb->cattrs);
+	if (!(succeeded = mono_image_add_cattrs (assembly, tb->table_idx, MONO_CUSTOM_ATTR_TYPEDEF, tb->cattrs, error)))
+		goto leave;
 	if (tb->fields) {
-		for (i = 0; i < tb->num_fields; ++i) {
+		for (i = 0; succeeded && i < tb->num_fields; ++i) {
```

Why can't we fail fast here? This code looks complicated.